kernel-crawler.py

```diff
@@ -1,461 +1,751 @@
+        for repo in repos:
+            for release, dependencies in repo.get_package_tree(version).items():
+                packages.setdefault(release, set()).update(dependencies)
+        return packages
+
+    def get_package_urls(self, version=''):
+        urls = []
+        for dependencies in self.get_package_tree(version).values():
+            urls.extend(dependencies)
+        return urls
+
+
+class MultiMirror(Mirror):
+    def __init__(self, mirrors):
+        self.mirrors = mirrors
+
+    def list_repos(self):
+        repos = []
+        for mirror in self.mirrors:
+            repos.extend(mirror.list_repos())
+        return repos
+
+
+# --- Debian-like repos ---
+
+
+class IncompletePackageListException(Exception):
+    pass
+
+
+class DebRepository(Repository):
+
+    def __init__(self, repo_base, repo_name):
+        self.repo_base = repo_base
+        self.repo_name = repo_name
+
+        dist = repo_name.replace('dists/', '').split('/')[0].split('-')[0]
+        self.dist = dist
+
+    def __str__(self):
+        return self.repo_base + self.repo_name
+
+    def is_valid(self):
+        return check_any([
+            self.repo_base + self.repo_name + '/Packages.gz',
+            self.repo_base + self.repo_name + '/Packages.xz',
+        ])
+
+    @classmethod
+    def scan_packages(cls, stream):
+        """
+        Parse a Packages file into individual packages metadata.
+        """
+        current_package = {}
+        packages = {}
+        for line in stream:
+            line = line.rstrip()
+            if line == '':
+                name = current_package['Package']
+                depends = current_package.get('Depends', [])
+                packages[name] = {
+                    'Depends': set(depends),
+                    'Version': current_package['Version'],
+                    'Filename': current_package['Filename'],
+                }
+                current_package = {}
+                continue
+            # ignore multiline values
+            if line.startswith(' '):
+                continue
             try:
-                progress(repo["root"], vid, len(versions), version)
-                version_page=urlopen(version_url,timeout=URL_TIMEOUT).read()
-            except:
+                key, value = line.split(': ', 1)
+                if key in ('Provides', 'Depends'):
+                    value = value.split(', ')
+            except ValueError:
+                print(line)
+                raise
+            current_package[key] = value
+
+        if current_package:
+            name = current_package['Package']
+            depends = current_package.get('Depends', [])
+            packages[name] = {
+                'Depends': set(depends),
+                'Version': current_package['Version'],
+                'Filename': current_package['Filename'],
+            }
+
+        return packages
+
+    KERNEL_PACKAGE_PATTERN = re.compile(r'^linux-.*?-[0-9]\.[0-9]+\.[0-9]+')
```

Comment:
```suggestion
    KERNEL_PACKAGE_PATTERN = re.compile(r'^linux-.*?-[0-9]+\.[0-9]+\.[0-9]+')
```
just to be future-proof :)

Reply:
I don't have an opinion on this. Do you believe this code will survive until Linux 10.0?

Reply:
It depends on whether they decide to change their versioning speed or not. Some years ago I supposed 2.x would have lived forever.
Same if you think about Firefox, we're about to hit version 100. I would never have imagined that!

Reply:
In some way that change has already happened after Linus got bored of 2.6.lots and we're up to 5.something now ;)